patches/server/1047-Add-FluidState-API.patch

```diff
@@ -0,0 +1,239 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: vicisacat <victor.branchu@gmail.com>
+Date: Fri, 17 Nov 2023 20:22:43 +0100
+Subject: [PATCH] Add FluidState API
+
+
+diff --git a/src/main/java/io/papermc/paper/block/fluid/PaperFluidData.java b/src/main/java/io/papermc/paper/block/fluid/PaperFluidData.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..1c7e4d2ce511e916226c6afabcd9be83edbda391
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/block/fluid/PaperFluidData.java
+@@ -0,0 +1,129 @@
++package io.papermc.paper.block.fluid;
++
++import com.google.common.base.Preconditions;
++import io.papermc.paper.block.fluid.type.PaperFlowing;
++import io.papermc.paper.block.fluid.type.PaperFallingFluid;
++import java.util.HashMap;
++import java.util.Map;
++import java.util.function.Function;
++import io.papermc.paper.util.MCUtil;
++import net.minecraft.core.BlockPos;
++import net.minecraft.world.level.block.state.properties.Property;
++import net.minecraft.world.level.material.FluidState;
++import net.minecraft.world.level.material.LavaFluid;
++import net.minecraft.world.level.material.WaterFluid;
++import org.bukkit.Fluid;
++import org.bukkit.Location;
++import org.bukkit.World;
++import org.bukkit.craftbukkit.CraftFluid;
++import org.bukkit.craftbukkit.CraftWorld;
++import org.bukkit.craftbukkit.util.CraftVector;
++import org.bukkit.util.Vector;
++
++public class PaperFluidData implements FluidData {
++
++    private static final Map<Class<? extends net.minecraft.world.level.material.Fluid>, Function<FluidState, PaperFluidData>> MAP = new HashMap<>();
++
++    private final FluidState state;
++
++    protected PaperFluidData(final FluidState state) {
++        this.state = state;
++    }
++
++    /* Registry */
++
++    static {
++        //<editor-fold desc="PaperFluidData Registration" defaultstate="collapsed">
++        register(LavaFluid.Source.class, PaperFallingFluid::new);
++        register(WaterFluid.Source.class, PaperFallingFluid::new);
++        register(LavaFluid.Flowing.class, PaperFlowing::new);
++        register(WaterFluid.Flowing.class, PaperFlowing::new);
```

Comment:
Where exactly does this handle air? Kinda confused.

Reply:
Didn't do that, this was in MachineMaker's original PR and to be honest I have no idea how it works. If i get the FluidData of an air block, and do getFluidType() it returns Fluid.EMPTY so uh it works i guess?

Reply:
Okay cool

Reply:
I didn't include empty here because it's handled below in the getOrDefault call on the map when creating an instance. the empty just uses the regular PaperFluidData type.